src/greenletio/core.py

```diff
@@ -77,24 +94,30 @@ async def main():
             await fn()
 
     :param fn: the standard function to convert to async.
+    :param copy_context: pass current contextvars.Context to greenlet (make available in fn)
+    :param copyback_context: reflect changes made in greenlet in current context
+        Using this option together with `copy_context` make this call similar to awaiting normal async function
     """
     @functools.wraps(fn)
-    def decorator(*args, **kwargs):
-        async def wrapper(fn, *args, **kwargs):
-            gl = greenlet(fn)
-            coro = gl.switch(*args, **kwargs)
-            while gl:
-                try:
-                    result = await coro
-                except:  # noqa: E722
-                    coro = gl.throw(*sys.exc_info())
-                else:
-                    coro = gl.switch(result)
-
-            return coro
-        return wrapper(fn, *args, **kwargs)
-
-    return decorator
+    async def wrapper(*args, **kwargs):
```

Comment:
It looks like this breaks the decorator usage. I don't understand why you have changed the structure of this decorator.

Reply:
Cause previously it return sync function which can't be checked by `asyncio.iscoroutinefunction`. It returns `False`, but function actually can be awaited

It does not break previous decorator usage. But it's not possible to pass params, yes.

We need add few lines to support parametrized decorator

```python
def async_(fn=None):
...

    if fn is None:
        return functools.partial(copy_context=copy_context, copyback_context=copyback_context)
    return wrapper
```

And after that we can use it:

```python

@async_
def func(): ...

@async_()
def func(): ...

@async_(copy_context=True)
def func(): ...
```